packages/react/src/components/Form/PinField.tsx

```diff
@@ -0,0 +1,157 @@
+  const fieldClassNames = ['form-field', state].join(' ');
+  const labelClassNames = ['form-label', state].join(' ');
+
+  const defaultValues = Array.from({ length }).map((_, i) => value?.toString()[i] || '');
+  const [values, setValues] = useState(defaultValues);
+  const fields = usePinHook({
+    values,
+    autoFocus,
+    secret,
+    pattern,
+    state,
+    onChange: v => setValues(v),
+    onComplete,
+  });
+
+  return (
+    <div className="form-field-row">
+      {label && (
+        <label className={labelClassNames}>
+          {label}
+        </label>
+      )}
+      <div className={containerClassNames}>
+        {fields.map((fieldProps, index) => (
+          <div key={index} className={fieldClassNames}>
+            <input className="form-input" {...fieldProps} />
+          </div>
+        ))}
+      </div>
+    </div>
+  );
+};
+
+PinField.displayName = 'PinField';
+
+const usePinHook = ({
+  values,
+  autoFocus,
+  secret,
+  state,
+  pattern,
+  onChange: onChangeProp,
+  onComplete,
+}: PinHookProps = {}) => {
+  const disabled = state === 'disabled';
+  const error = state === 'error';
+
+  const fieldRefs = useRef<HTMLInputElement[]>(Array(values.length).fill(null));
+
+  const setFocus = useCallback((index = 0) => {
+    fieldRefs.current[index]?.focus();
+  }, []);
+
+  useEffect(() => {
+    if (autoFocus) setFocus();
+  }, [autoFocus, setFocus]);
+
+  const setFieldRef = useCallback((index: number) => (ref: HTMLInputElement) => {
+    fieldRefs.current[index] = ref;
+  }, []);
+
+  const onChange = useCallback((index: number) => (event: ChangeEvent<HTMLInputElement>) => {
+    let { value } = event.target;
+
+    value = value.trim().toUpperCase();
+
+    if (!pattern.test(value)) return;
+
+    const nextValues = values.slice();
+    nextValues[index] = value;
+    onChangeProp(nextValues);
+
+    if (value) {
+      if (!nextValues.includes('')) {
+        onComplete?.(nextValues.join(''));
+      }
+
+      if (index !== values.length - 1) {
+        if (error) {
+          const emptyFieldIndex = nextValues.findIndex((v) => !v);
+
+          if (emptyFieldIndex !== -1) {
+            setFocus(emptyFieldIndex);
+          }
+        } else {
+          setFocus(index + 1);
+        }
+      }
+    }
+  }, [pattern, values, onChangeProp, onComplete, error, setFocus]);
+
+  const onKeyDown = useCallback((index: number) => (event: KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Backspace' && !values[index] && index) {
+      setFocus(index - 1);
+    }
+  },[values, setFocus]);
+
+  const fields = values.map((value: string, index: number) => ({
+    ref: setFieldRef(index),
+    value,
+    disabled,
+    autoComplete: 'off',
+    maxLength: 1,
+    type: secret ? 'password' : 'text',
+    ...(!disabled && {
+      onChange: onChange(index),
+      onKeyDown: onKeyDown(index),
+    }),
+  }));
+
+  return fields;
+};
+
+const DEFAULT_PATTERN = /^[a-zd]*$/i;
```

Comment:
This pattern does not allow number, but the letters a-z and `d` 

```suggestion
const DEFAULT_PATTERN = /^[a-z0-9]*$/i;
```

---

Should there be a pattern at all? 
- it just adds complexity
- restricting the user to a pattern is a use case when creating a pin, but when logging in with existing pin there's no need to apply a pattern

Instead of a `DEFAULT_PATTERN` of letters/numbers it might be best to have no pattern or `/*/` by default

Reply:
I guess you mean `/./`? I don't think `/*/` is a valid pattern.

Reply:
Yes, though I think to be able to delete (backspace) you'd need to have `/.*/`